homeassistant/components/homekit_controller/device_trigger.py

```diff
@@ -0,0 +1,269 @@
+    # manufacturer might not - clamp options to what they say.
+    all_values = clamp_enum_to_char(InputEventValues, input_event)
+
+    results = []
+    for event_type in all_values:
+        results.append(
+            {
+                "characteristic": input_event.iid,
+                "value": event_type,
+                "type": "doorbell",
+                "subtype": HK_TO_HA_INPUT_EVENT_VALUES[event_type],
+            }
+        )
+    return results
+
+
+TRIGGER_TYPES = {
+    "service-label": enumerate_stateless_switch_group,
+    "stateless-programmable-switch": enumerate_stateless_switch,
+    "doorbell": enumerate_doorbell,
+}
+
+
+async def async_setup_triggers_for_entry(hass: HomeAssistant, config_entry):
+    """Triggers aren't entities as they have no state, but we still need to set them up for a config entry."""
+    hkid = config_entry.data["AccessoryPairingID"]
+    conn = hass.data[KNOWN_DEVICES][hkid]
+
+    @callback
+    def async_add_service(aid, service_dict):
+        service_type = service_dict["stype"]
+
+        # If not a known service type then we can't handle any stateless events for it
+        if service_type not in TRIGGER_TYPES:
+            return False
+
+        # We can't have multiple trigger sources for the same device id
+        # Can't have a doorbell and a remote control in the same accessory
+        # They have to be different accessories (they can be on the same bridge)
+        # In practice, this is inline with what iOS actually supports AFAWCT.
+        device_id = conn.devices[aid]
+        if device_id in hass.data[TRIGGERS]:
+            return False
+
+        # At the moment add_listener calls us with the raw service dict, rather than
+        # a service model. So turn it into a service ourselves.
+        accessory = conn.entity_map.aid(aid)
+        service = accessory.services.iid(service_dict["iid"])
+
+        # Just because we recognise the service type doesn't mean we can actually
+        # extract any triggers - so only proceed if we can
+        triggers = TRIGGER_TYPES[service_type](service)
+        if len(triggers) == 0:
+            return False
+
+        trigger = TriggerSource(conn, aid, triggers)
+        hass.data[TRIGGERS][device_id] = trigger
+
+        return True
+
+    conn.add_listener(async_add_service)
+
+
+def async_fire_triggers(conn, events):
+    """Process events generated by a HomeKit accessory into automation triggers."""
+    for (aid, iid), ev in events.items():
+        if aid not in conn.devices:
+            continue
+        device_id = conn.devices[aid]
+        if device_id in conn.hass.data[TRIGGERS]:
+            source = conn.hass.data[TRIGGERS][device_id]
+            source.fire(iid, ev)
+
+
+async def async_get_triggers(hass: HomeAssistant, device_id: str) -> List[dict]:
+    """List device triggers for homekit devices."""
+
+    if device_id not in hass.data[TRIGGERS]:
```

Comment:
Sometimes the device triggers are not loaded and I noticed the following exception:

```
hass    | 2020-08-24 23:40:47 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.140188964756160] Error handling message: Unknown error
hass    | Traceback (most recent call last):
hass    |   File "/usr/src/homeassistant/homeassistant/components/websocket_api/decorators.py", line 21, in _handle_async_response
hass    |     await func(hass, connection, msg)
hass    |   File "/usr/src/homeassistant/homeassistant/components/device_automation/__init__.py", line 188, in with_error_handling
hass    |     await func(hass, connection, msg)
hass    |   File "/usr/src/homeassistant/homeassistant/components/device_automation/__init__.py", line 238, in websocket_device_automation_list_triggers
hass    |     triggers = await _async_get_device_automations(hass, "trigger", device_id)
hass    |   File "/usr/src/homeassistant/homeassistant/components/device_automation/__init__.py", line 134, in _async_get_device_automations
hass    |     device_automations = await asyncio.gather(
hass    |   File "/usr/src/homeassistant/homeassistant/components/device_automation/__init__.py", line 109, in _async_get_device_automations_from_domain
hass    |     return await getattr(platform, function_name)(hass, device_id)
hass    |   File "/usr/src/homeassistant/homeassistant/components/homekit_controller/device_trigger.py", line 237, in async_get_triggers
hass    |     if device_id not in hass.data[TRIGGERS]:
hass    | KeyError: 'homekit_controller-triggers'
```

Seems it is caused by a race condition when the initialisation is not fully done. Not sure how to wait for the initialisation though... or just set a default value?
```suggestion
    device = hass.data[TRIGGERS].get(device_id, {})
```

Reply:
Does that mean its trying to load the automations before the config entry is ready? If triggers are being set up before the config entry is loaded then the triggers code would fail because the entity map won't have been parsed and the triggers objects created right? So we need to find a way to make sure these properly wait for the device to be ready.

Even listing available triggers would require the config entry to have been loaded, which i thought meant that async_setup should have finished being invoked?

Reply:
> Does that mean its trying to load the automations before the config entry is ready? 
Yes, I think so. I think last time I saw it is when I was restarting HA with automation configuration UI open. But that doesn't alway happen... need to watch this issue.

Reply:
I wonder if you stick a await asyncio.sleep(60000000) in async_setup of homekit_controller if you can then reproduce this reliably?

Reply:
Yes, adding `await asyncio.sleep(60000000)` to `async_setup` and open the device automation setup page reproduces that. It might not be a big issue since after the config entry is loaded, the device automation setup page will work again. A simple fix might be just providing a default value `{}`.

Reply:
Yeah I think I agree. I've pushed your suggested fix.